classes/DependencyInjection/ContainerProvider.php

```diff
@@ -0,0 +1,90 @@
+<?php
+/**
+ * 2007-2020 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Academic Free License 3.0 (AFL-3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/AFL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/AFL-3.0 Academic Free License 3.0 (AFL-3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\Module\PrestashopCheckout\DependencyInjection;
+
+use PrestaShop\Module\PrestashopCheckout\Cache\CachePathProviderInterface;
+use PrestaShop\Module\PrestashopCheckout\Configuration\Configuration;
+use Symfony\Component\Config\ConfigCache;
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\DependencyInjection\Dumper\PhpDumper;
+use Symfony\Component\DependencyInjection\Loader\YamlFileLoader;
+
+class ContainerProvider
+{
+    /**
+     * @var Configuration
+     */
+    private $configuration;
+
+    /**
+     * @var CachePathProviderInterface
+     */
+    private $cachePathProvider;
+
+    /**
+     * @param Configuration $configuration
+     * @param CachePathProviderInterface $cachePathProvider
+     */
+    public function __construct(Configuration $configuration, CachePathProviderInterface $cachePathProvider)
+    {
+        $this->configuration = $configuration;
+        $this->cachePathProvider = $cachePathProvider;
+    }
+
+    /**
+     * @param string $containerName
+     *
+     * @return ContainerInterface
+     */
+    public function getContainer($containerName)
+    {
+        $containerClassName = ucfirst($this->configuration->getModuleName())
+            . (false === empty($containerName) ? ucfirst($containerName) : '')
+            . 'Container'
+        ;
+        $containerFilePath = $this->cachePathProvider->getPath() . '/' . $containerClassName . '.php';
+        $containerConfigCache = new ConfigCache($containerFilePath, $this->configuration->isDebug());
+
+        if ($containerConfigCache->isFresh()) {
+            require_once $containerFilePath;
+
+            return new $containerClassName();
```

Comment:
I see that you're using the class `Configuration`, and define it in the services afterwards. I wonder if this class won't be instantiated two times.
This seems acceptable today as the class do not execute heavy tasks in its constructor, but we may encounter issues later with further changes.

We could avoid instantiation of the same class by affecting the instance we already have in hand:

```php
$containerBuilder->set('configuration', $this->configuration);
```

Reply:
Applicable only if we build our own container, if we are on PrestaShop 1.7.3+ Core container will not register them...

Reply:
Of course, keep its definition in your services.yml files, here is a a shortcut to reuse an existing instance instead of creating another one.